### ISSUE TITLE
Reschedule all cards works differently between the FSRS4 Javascript (Anki 2.1.66) and the FSRS4 native (Anki 23.10) [BUG]

### ISSUE THREAD

**Issue by cjdduarte** [open-spaced-repetition/fsrs4anki-helper#253]
**Describe the bug**

With the same weights in both versions and using the same database, when triggering the "Reschedule all cards" function in Anki 2.1.66 for today, there are 175 cards to review.

In Anki 23.10 using the integrated FSRS, and again triggering the "Reschedule all cards" function, it tells me to review 325 cards.

**Environment**
 - Anki version: [e.g. 2.1.66 and 23.10]

**Comment by L-M-Sherlock** [comment]
Could you check the extra reviews' card info?

**Comment by cjdduarte** [comment]
**Anki 2.1.66**

![image](https://github.com/open-spaced-repetition/fsrs4anki-helper/assets/42243532/0c60898c-4e16-49d8-aaa7-08b4e17d956f)

**Anki 23.10**

![image](https://github.com/open-spaced-repetition/fsrs4anki-helper/assets/42243532/23bfb0d2-e63d-4569-b2df-6f3cbf4b6059)

**Comment by user1823** [comment]
It seems to be the same issue as https://forums.ankiweb.net/t/anki-23-10-beta/34912/87.

In v23.10, the add-on uses Anki's FSRS code to calculate the stability.

**Comment by cjdduarte** [comment]
> It seems to be the same issue as https://forums.ankiweb.net/t/anki-23-10-beta/34912/87.
> 
> In v23.10, the add-on uses Anki's FSRS code to calculate the stability.

So in this case, would Anki 23.10 be "more correct"?
Because this has been my perception. Or is it the opposite?

**Comment by cjdduarte** [comment]
0.3524, 0.3869, 1.9765, 22.3079, 4.7508, 1.0115, 0.8749, 0.0000, 1.5617, 0.1000, 1.0117, 2.1119, 0.1170, 0.2724, 1.2289, 0.0626, 2.6100

**Comment by user1823** [comment]
And your request/desired retention?

By the way, the stability calculated by the helper add-on (in 2.1.66) is correct based on the weights.

**Comment by cjdduarte** [comment]
> And your request/desired retention?
> 
> By the way, the stability calculated by the helper add-on (in 2.1.66) is correct based on the weights while that calculated by Anki is not.

0.9

**Comment by user1823** [comment]
> > And your request/desired retention?
> 
> 0.9

Fine. So, the helper add-on (in 2.1.66) is working correctly but the built-in FSRS in Anki 23.10 is not. @L-M-Sherlock, this needs to be investigated.

@cjdduarte, if you think that the intervals given by the add-on are incorrect, it might be possible that the optimizer has some bug. But before suspecting that, read this:

> For many users, SM-2 tends to show new cards at unnecessarily short intervals. So, when users switch to FSRS, they tend to feel that the intervals for new cards are too large. But these larger intervals match the target retention (configured in the scheduler code) better. By using these larger intervals, FSRS can prevent many of the unnecessary reviews that happen when using SM-2.
> 
> If you still want to decrease the intervals, you can increase the requestRetention in the scheduler. But note that this will decrease all the intervals, not just the first intervals.

**Comment by L-M-Sherlock** [comment]
Yeah, the result of built-in FSRS is incorrect. I will try to reproduce this bug.

**Comment by cjdduarte** [comment]
> > **Anki 23.10**
> 
> Could you move the card into a separate deck and reschedule the deck? I can't reproduce this bug in my device. Or you can send your collection file to me, and it will help me locate the bug quickly.

[07.5-Segurança da Informação.apkg.zip](https://github.com/open-spaced-repetition/fsrs4anki-helper/files/12767487/07.5-Seguranca.da.Informacao.apkg.zip)

I took that example of mine. See if it helps

**Comment by cjdduarte** [comment]
0.3524, 0.3869, 1.9765, 22.3079, 4.7508, 1.0115, 0.8749, 0.0000, 1.5617, 0.1000, 1.0117, 2.1119, 0.1170, 0.2724, 1.2289, 0.0626, 2.6100

**Comment by L-M-Sherlock** [comment]
<img width="827" alt="image" src="https://github.com/open-spaced-repetition/fsrs4anki-helper/assets/32575846/b8ff4027-db86-4914-a402-776e747af5d2">

It's correct in my device after rescheduling.

**Comment by cjdduarte** [comment]
> <img alt="image" width="827" src="https://user-images.githubusercontent.com/32575846/271635998-b8ff4027-db86-4914-a402-776e747af5d2.png">
> It's correct in my device after rescheduling.

Anki 2.1.66 or Anki 23.10 (beta3)?

This weight?: 0.3524, 0.3869, 1.9765, 22.3079, 4.7508, 1.0115, 0.8749, 0.0000, 1.5617, 0.1000, 1.0117, 2.1119, 0.1170, 0.2724, 1.2289, 0.0626, 2.6100

**Comment by L-M-Sherlock** [comment]
Version ⁨23.10 (8214aa42)⁩, with your weights.

**Comment by cjdduarte** [comment]
I just tested it again here. Check this out:
My Anki is the Version ⁨23.10 (d2f392ff)⁩

![image](https://github.com/open-spaced-repetition/fsrs4anki-helper/assets/42243532/a64e6413-74ab-4c7c-904f-4b3487950318)

**Comment by cjdduarte** [comment]
Something very strange happened:
In my original deck, it persists at that value. But when I exported the card and imported it again, the value corrected.

![image](https://github.com/open-spaced-repetition/fsrs4anki-helper/assets/42243532/443b641b-283f-4c17-b6a9-7cf2b2dd69be)

**Comment by L-M-Sherlock** [comment]
I guess the memory state didn't update normally in your first attempt.

**Comment by cjdduarte** [comment]
I managed to reproduce it:

1 - Import this deck:

[collection-2023-09-29@14-01-26.zip](https://github.com/open-spaced-repetition/fsrs4anki-helper/files/12769397/collection-2023-09-29%4014-01-26.zip)

2 - Delete the Javascript, enable the SRFS, and input these weights: 0.3524, 0.3869, 1.9765, 22.3079, 4.7508, 1.0115, 0.8749, 0.0000, 1.5617, 0.1000, 1.0117, 2.1119, 0.1170, 0.2724, 1.2289, 0.0626, 2.6100
3 - Save
4 - Order to reschedule all
5 - Filter by this ID: 2261631

![image](https://github.com/open-spaced-repetition/fsrs4anki-helper/assets/42243532/07b1ea6a-1754-44e0-bcc9-111cb935679e)

**Comment by cjdduarte** [comment]
However, if I export just this card, delete it, and then reimport it, it goes back to those above values. 

![image](https://github.com/open-spaced-repetition/fsrs4anki-helper/assets/42243532/25097b36-1166-46a3-a322-d7df95bf518e)

But the important thing is that I managed to reproduce the error with the steps above.

**Comment by cjdduarte** [comment]
I FOUND OUT.

I have several deck groups. When applying the weight, it is only saving in the current one.

![image](https://github.com/open-spaced-repetition/fsrs4anki-helper/assets/42243532/a5ac98dc-d3c6-4be4-a024-cf4b88620bc4)

I had to command it to save in all.

![image](https://github.com/open-spaced-repetition/fsrs4anki-helper/assets/42243532/88c9a5b4-eab0-4941-9d41-6198f95ce948)

I think many people are not aware of this.

**Comment by cjdduarte** [comment]
Is the standard to apply the weights by group or generally?

**Comment by L-M-Sherlock** [comment]
The weights only works for those decks using the preset. It will use the default weights if the weights field is empty.

**Comment by L-M-Sherlock** [comment]
@dae, we need to write a document for that. Where can we write it?

**Comment by cjdduarte** [comment]
> The weights only works for those decks using the preset. It will use the default weights if the weights field is empty.

Resolved. All the differences were because of this.
The problem is that when no weight exists, the Helper doesn't run. However, when I insert 1, it enables.

Suggestion: As you said yourself, either make this very well informed, or also make it possible to be universal like "Custom Scheduler".

**Comment by dae** [comment]
@L-M-Sherlock I will add documentation inside the decks screen when I have a chance

@cjdduarte If all of your decks have approximately the same difficulty, you can make them share the one preset if you want a single set of weights.

**Comment by cjdduarte** [comment]
> @L-M-Sherlock I will add documentation inside the decks screen when I have a chance
> 
> @cjdduarte If all of your decks have approximately the same difficulty, you can make them share the one preset if you want a single set of weights.

Here's the translated text:

@dae , I'm still thinking about how I will reorganize my decks.
As I study to take exams and the content is extensive, I apply "weights" to the subjects.

For example, if one subject is worth 30% of the exam grade and another subject is worth 5% of the exam, I set up to study 30 new cards from that subject and 5 cards from another subject. Thus, I maintain a 30/5 ratio.
Imagine this for 25 subjects (that's why I have several deck groups).

However, this solution becomes complicated with this new way of using the integrated FSRS.
I value simplicity. In my case, it would be registering the weights just once. Or simplifying the groups.

Anyway, I have to rethink the best way between how I have studied today and the new functionality. I have to adapt to the new way of using Anki.

**Comment by user1823** [comment]
In the v3 scheduler, the new cards per day setting can be deck-specific in addition to being preset-specific.

**Comment by cjdduarte** [comment]
> In the v3 scheduler, the new cards per day setting can be deck-specific in addition to being preset-specific.

@user1823 , good point. I know about the feature, but I've never used it. As a long-time user, sometimes we get stuck in our old ways.

It will probably be the most correct solution for my case.
Thanks for reminding me.

**Comment by user1823** [comment]
> > In the v3 scheduler, the new cards per day setting can be deck-specific in addition to being preset-specific.
> 
> @user1823 , good point. I know about the feature, but I've never used it. As a long-time user, sometimes we get stuck in our old ways.
> 
> It will probably be the most correct solution for my case. Thanks for reminding me.

Well, the same case was with me. Just a few days ago, Damien reminded me of the feature.

**Comment by cjdduarte** [comment]
@dae  and @L-M-Sherlock , I noticed that the log functionality was removed from the native version. 
However, wouldn’t it be the case to keep it and add the information about the current card weight?
Example:

```
FSRS enabled
**Preset: (Default or XYZ or ABC)**
D: 10
S: 5.74
R: 89.59%
```

I think this would help those who want to keep the logs to learn what's happening in this improvement?

![image](https://github.com/open-spaced-repetition/fsrs4anki-helper/assets/42243532/80205119-ff64-4241-bee5-7431ee1e2222)
